feat: add trackio as experiment tracking integration

abidlabs opened the pull request:
## Summary
This PR adds trackio as a new experiment tracking option in axolotl, similar to wandb, mlflow, and comet. ([Trackio](https://github.com/gradio-app/trackio) is a local-first, free, open-source experiment tracking library by Hugging Face).

How it works:
- Trackio runs **locally by default** when `trackio_project_name` is set
- Optionally syncs to a Hugging Face Space if `trackio_space_id` is provided
- Integrates natively with transformers via `report_to`

Usage: just set:
```yaml
trackio_project_name: "my-project"
trackio_run_name: "experiment-1"
trackio_space_id: "username/my-space"  # optional
```



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added TrackIO integration support for experiment tracking and configuration management.
  * New configuration options: trackio_project_name, trackio_run_name, and trackio_space_id for flexible TrackIO setup.
  * System now automatically saves configurations to TrackIO during training.

* **Chores**
  * Updated TrackIO dependency to version 0.11.0 or later.
  * Updated example configurations to include new TrackIO settings.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

abidlabs committed:
feat: add trackio as experiment tracking integration

- Add TrackioConfig to integrations schema with project_name, run_name, and space_id
- Create trackio_.py module for environment setup
- Add is_trackio_available() utility function
- Integrate trackio with report_to in trainer builder
- Add trackio callback for experiment tracking
- Add trackio config keys to gpt-oss example YAMLs
- Trackio runs locally by default, syncs to HF Space if space_id provided

NanoCode012 commented on src/axolotl/utils/callbacks/trackio_.py:
```diff
@@ -0,0 +1,32 @@
+from axolotl.utils.logging import get_logger
+
+if TYPE_CHECKING:
+    from axolotl.core.training_args import AxolotlTrainingArguments
+
+LOG = get_logger(__name__)
+
+
+class SaveAxolotlConfigtoTrackioCallback(TrainerCallback):
+    """Callback for trackio integration"""
+
+    def __init__(self, axolotl_config_path):
+        self.axolotl_config_path = axolotl_config_path
+
+    def on_train_begin(
+        self,
+        args: "AxolotlTrainingArguments",
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        if is_main_process():
+            LOG.info("Trackio experiment tracking is enabled.")
+        return control
```
Just in case missed, we would upload the config here.

Will a user need to have the local server running to accept file uploads?

abidlabs replied:
Fixed now in https://github.com/axolotl-ai-cloud/axolotl/pull/3253/commits/5a0291f6c9e61e7b0f37b4675e64ebcf677b467c. Yes, the user will need to the hvave local server running, but this automatically taken care of when a trackio project is initialized with `trackio.init`.

winglian commented on src/axolotl/utils/schemas/integrations.py:
```diff
@@ -200,3 +200,23 @@ class OpenTelemetryConfig(BaseModel):
     )
+
+
+class TrackioConfig(BaseModel):
```
This also needs to be added to https://github.com/axolotl-ai-cloud/axolotl/blob/main/src/axolotl/utils/schemas/config.py#L54

abidlabs replied:
Addressed in https://github.com/axolotl-ai-cloud/axolotl/pull/3253/commits/333054bbe81abb759267866baafeba139252054b

contrastive-vae committed:
changes

contrastive-vae committed:
changes

contrastive-vae committed:
changes

abidlabs commented on requirements.txt:
```diff
@@ -20,7 +20,7 @@ deepspeed>=0.17.0
 hf_xet==1.2.0
 kernels>=0.9.0
-trackio
+trackio>=0.11.0
```
This is incorrect, see https://pypi.org/project/trackio/

contrastive-vae committed:
changes

contrastive-vae committed:
changes

abidlabs commented on requirements.txt:
```diff
@@ -20,12 +20,12 @@ deepspeed>=0.17.0
 hf_transfer
 sentencepiece
-gradio==5.49.1
+gradio>=6.0,<7.0
```
While I was at it, I upgraded the version of `gradio` to 6.0, as this is compatible with the latest version of `trackio`, and brings other benefits, e.g. Gradio 6 .whl is much smaller than the Gradio 5 .whl. I went through the code and updated the Gradio app to use Gradio 6 syntax.

NanoCode012 replied:
Would you be able to test that Gradio features still work?

abidlabs replied:
Sure, happy to test and confirm

contrastive-vae committed:
changes

NanoCode012 commented:
Thanks for the update, letting tests run

abidlabs commented:
Please let me know if the test failures are expected (they don't seem related to my changes but I could be mistaken) @NanoCode012

NanoCode012 commented:
@abidlabs , no these ones are new. Not sure if they are from the newer requirements. I'll let CI run again to be sure,

NanoCode012 commented:
@abidlabs , the CI met the same issue again across 3 env. I don't see this happening on other PRs. Could you check if it's from trackio update/gradio update?

abidlabs commented:
Thanks @NanoCode012 for confirming I’ll look into this

abidlabs committed:
changes

abidlabs commented:
Hi @NanoCode012 could we run the CI again? I updated the version of the `typing_extensions` dependency. 

I also confirmed that the Gradio integration continues to work with Gradio 6. Here's the UI, launched with: `axolotl inference config.yml --lora-model-dir="./outputs/lora-out" --gradio`

<img width="1307" height="325" alt="image" src="https://github.com/user-attachments/assets/54aa8733-2dae-46bb-875c-141165ae4444" />


And here's the config yml I used:

```yml
base_model: google/gemma-3-270m

model_type: Gemma3ForCausalLM

# Automatically upload checkpoint and final model to HF
# hub_model_id: username/custom_model_name

load_in_8bit: false
load_in_4bit: true

datasets:
  - path: mhenrichsen/alpaca_2k_test
    type: alpaca
    split: train[:5]

val_set_size: 0.0
output_dir: ./outputs/lora-out

adapter: qlora
lora_model_dir:

sequence_len: 256
sample_packing: false
eval_sample_packing: false


lora_r: 4
lora_alpha: 8
lora_dropout: 0.0
lora_target_linear: true
lora_mlp_kernel: false
lora_o_kernel: false
lora_qkv_kernel: false

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 1
micro_batch_size: 1
max_steps: 3

optimizer: adamw_bnb_8bit
lr_scheduler: constant
learning_rate: 0.0002

bf16: auto
tf32: true

gradient_checkpointing: false
resume_from_checkpoint:
logging_steps: 5
flash_attention: false

warmup_ratio: 0.0
evals_per_epoch: 0
save_strategy: "steps"
save_steps: 3
weight_decay: 0.0
special_tokens:
  pad_token: "<|end_of_text|>"

# save_first_step: true  # uncomment this to validate checkpoint saving works with your config
```

abidlabs commented on requirements.txt:
```diff
@@ -20,12 +20,13 @@ deepspeed>=0.17.0
 hf_transfer
 sentencepiece
-gradio==5.49.1
+gradio>=6.0,<7.0
```
```suggestion
gradio>=6.2.0,<7.0
```

abidlabs committed:
Update requirements.txt

abidlabs commented:
Hi @NanoCode012 is it possible to rerun the CI? We just released a new version of Gradio that relaxes the pydantic version (https://github.com/gradio-app/gradio/pull/12598), which may fix the issue that we're seeing in the CI runs

winglian commented:
I'm not worried about the pytest ones failing since that's a known issue that is fixed in main. once the integration tests pass, we're good to merge

winglian committed:
don't allow pydantic 2.12 for now

abidlabs commented:
Got it, thanks @NanoCode012, it looks like we are good to merge now?
